src/modules/sync-cli.js

```diff
@@ -781,9 +762,83 @@ var self = {
     });
   },
 
-  existingPendingPreHashMap: function(dataset){
-    var pendingPreHashes = dataset.changeHistory || {};
-    return pendingPreHashes;
+  applyPendingChangesToRecords: function(dataset, records){
+    var pendings = dataset.pending;
+    for(var pendingUid in pendings){
+      if(pendings.hasOwnProperty(pendingUid)){
+        var pendingObj = pendings[pendingUid];
+        var uid = pendingObj.uid;
+        //if the records contain any thing about the data records that are currently in pendings,
+        //it means there are local changes that haven't been applied to the cloud yet,
+        //so update the pre value of each pending record to relect the latest status from cloud
+        //and remove them from the response
+        if(records.create){
+          var creates = records.create;
+          if(creates && creates[uid]){
+            delete creates[uid];
```

Comment:
@corinnekrych @edewit @secondsun 
Small changes here guys. We don't set the pre value of the pending records to be the same as the current cloud value, we just delete them from the response. The problem of the previous implementation is that it will not report collisions when it should be.

For example, local change from A to B, while that is submitting, the value is changed from B to C. Between the first request is completed and the second request is completed, there is another change from another client to change the value from B to D. So the response got back from the cloud will be: change value from C to D.

In the previous implementation, we are replacing the pending change's pre value to D (which means change from B-C to D-C), which means the submission will succeed and it will change the cloud value from D to C. However, this should be a conflict: one client is change the value from B to C and the other is changing the client from B to D.

So we should just ignore those changes if we have pending changes locally, and let the cloud decide if the change can be applied.